### ISSUE TITLE
Unused $event parameters in AOT generated code

### ISSUE THREAD

**Issue by nicolashenry** [angular/angular#17131]
**I'm submitting a ...**
```
[x] bug report
```

**Current behavior**
Running ngc generates functions that takes the event as argument. If the component template does not use the event and noUnusedParameters is set to true, this causes an error : 
`'$event' is declared but never used.`

**Expected behavior**
No error

**Minimal reproduction of the problem with instructions**
Just use aot with noUnusedParameters and a template listening an event but not using the $event variable (`<button (click)="doSomething()">Do</button>` for example)

**What is the motivation / use case for changing the behavior?**
Using noUnusedParameters with aot since https://github.com/angular/angular/issues/15532 was supposed to be fixed.

* **Angular version:** 4.2.0-rc.1

* **Browser:** all
 
* **Language:** TypeScript 2.3.2

* **Node (for AoT issues):** `node --version` =  v7.10.0

**Comment by Martin-Wegner** [comment]
Same problem with 4.2.0 :(

**Comment by tsvetomir** [comment]
This issue might not seem serious, but it will break applications that used to run fine before. The workarounds are not pretty either.

* You can switch _noUnusedParameters_ off. This has the potential of backfiring whenever you decide to turn it on back again.
* Add (ahem, _unused_) parameters to the event handler declaration _and_ implementation. It is also a thing you must remember to remove at some point.

**Comment by vatsalpande** [comment]
I am also facing the same issue. Normal build works just fine. But with aot .. tons of issues.
even from node modules folder. Any suggestions?

**Comment by ajafff** [comment]
My workaround: disable `noUnusedParameters` in tsconfig.json and add `no-unused` https://github.com/ajafff/tslint-consistent-codestyle/blob/master/docs/no-unused.md to `tslint.json`

That lint rule basically does the same as typescript's `noUnusedParameters`

**Comment by brunolm** [comment]
Explaning how to use @ajafff 's alternative:

- Install `npm i -D tslint-consistent-codestyle`
- On the top of your tslint file add

```
"rulesDirectory": [
  "tslint-consistent-codestyle"
],
```

On your tslint rules add

```
"no-unused": true
```

Remove from `tsconfig.json`

```
"noUnusedParameters": true
```

**Comment by Martin-Wegner** [comment]
Still present in Angular 4.3.4 and Angular CLI 1.3.0 :(

**Comment by sgwatgit** [comment]
If you prefix your parameter with `_` then the Typescript compiler doesn't complain.

It's a matter of taste, but I don't think this is a bad approach because the `_` tells developers this is a parameter my caller is giving me but I'm not using it. It also makes it easy to debug its value if you want.

`app.component.ts`
```
export class AppComponent {
  submitRequest(_$event: MouseEvent) {
    console.log('Ok!');
  }
}
```

`app.component.html`
```
<input type="submit" (click)="submitRequest($event)" value="Submit">
```
`my console`
```
$ ng build --prod
Date: 2017-08-18T19:03:35.269Z
Hash: 4a224d44f6c9a8db3b4e
Time: 13396ms
chunk {0} polyfills.ab340e00710313c68bfb.bundle.js (polyfills) 139 kB {4} [initial] [rendered]
chunk {1} main.bead538bb8f5f3c1abbe.bundle.js (main) 18.9 kB {3} [initial] [rendered]
chunk {2} styles.d41d8cd98f00b204e980.bundle.css (styles) 0 bytes {4} [initial] [rendered]
chunk {3} vendor.49160ce77bdc13855b0b.bundle.js (vendor) 310 kB [initial] [rendered]
chunk {4} inline.587ea3d0cd29db30f247.bundle.js (inline) 1.45 kB [entry] [rendered]
```

**Comment by vatsalpande** [comment]
The problem is with node modules . The compiler complains about that too

**Comment by jrmcdona** [comment]
Present in 4.3.6 and CLI 1.3.2.

**Comment by Martin-Wegner** [comment]
Still present in Angular 5.0.0-beta.7 with Angular CLI 1.5.0-beta.1 :(

**Comment by Martin-Wegner** [comment]
Seems to be fixed in 5.0.0-rc.0 :)

**Comment by nicolashenry** [comment]
Seems to be fixed in 5.0.0, I close this issue
